lib/shinq/launcher.rb

```diff
@@ -50,5 +55,16 @@ def lifecycle_limit?
       return false unless Shinq.configuration.lifecycle
       return (Shinq.configuration.lifecycle < @loop_count)
     end
+
+    private
+
+    def format_error_message(error)
+      if defined?(::Rails) && ::Rails.backtrace_cleaner
+        backtrace = ::Rails.backtrace_cleaner.clean(error.backtrace || [])
```

Comment:
Actually, backtrace_cleaner does not stop us from knowing the hierarchy. Rather, it helps us understand the backtrace more quickly, which contributes to debug efficiency.

By default, `backtrace_cleaner` does the following 2 operations.

* Filter: Show relative paths from application root instead of full paths from root `/`.
* Silencer: Exclude lines which belong to internal libraries.

http://api.rubyonrails.org/classes/ActiveSupport/BacktraceCleaner.html

The former does not break hierarchy at all. It just removes redundant information like `/home/app-user/application/releases/20170929123456/` from each line in backtrace. This significantly reduces noisy and redundant log messages especially on production servers.

The latter one also does not stop us from understanding the hierarchy. Since it removes line belonging to external libraries, the users' ruby codes in backtrace, from which almost all the errors are stemming, are not removed nor disordered. On the contrary, removing irrelevant lines makes the log concise and lets users focus on the real cause. In our environment, typical backtrace in worker has 16 lines in `serverengine`, 5 lines in `shinq`, and 15 lines in `bundler` compared with several (mainly 1 to 3) lines in our code, which means more than 90 % of backtrace consists of irrelevant lines.

Moreover, settings of backtrace_cleaner is usually maintained with application code typically under `config/initializers/backtrace_cleaners.rb`. This means that the filters and silencers are appropriately customized by each user, and in case whole backtrace are required, a user can see it just by calling `BacktraceCleaner#remove_filters!`.

Reply:
Just in case error occurs in external libraries, I will fix the code to show full backtrace when cleaned backtrace is empty.